pkg/plugins/crypto/crypto.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+// Package crypto provides the default implementation of pkg/framework/spi/crypto.Crypto interface
+//
+// It uses github.com/tink/go crypto primitives
+package crypto
+
+import (
+	"errors"
+
+	"github.com/google/tink/go/aead"
+	"github.com/google/tink/go/core/primitiveset"
+	"github.com/google/tink/go/keyset"
+	aeadsubtle "github.com/google/tink/go/subtle/aead"
+	"golang.org/x/crypto/chacha20poly1305"
+
+	cryptoapi "github.com/hyperledger/aries-framework-go/pkg/framework/spi/crypto"
+)
+
+// Crypto is the default Crypto implementation
+type Crypto struct {
+}
+
+// Assert that Crypto implements the Crypto interface.
+var _ cryptoapi.Crypto = (*Crypto)(nil)
+
+// Encrypt will encrypt msg using the implementation's corresponding encryption key
+func (t *Crypto) Encrypt(msg, aad []byte, kh *keyset.Handle) ([]byte, []byte, error) {
+	a, err := aead.New(kh)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ct, err := a.Encrypt(msg, aad)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ps, err := kh.Primitives()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ivSize := nonceSize(ps)
+	prefixLength := len(ps.Primary.Prefix)
+	cipherText := ct[prefixLength+ivSize:]
+	nonce := ct[prefixLength : prefixLength+ivSize]
+
+	return cipherText, nonce, nil
+}
+
+func nonceSize(ps *primitiveset.PrimitiveSet) int {
+	ivSize := aeadsubtle.AESGCMIVSize
+	// TODO double check if the right primitive from PrimitiveSet is Primary, or do we have to search for something else
+	if _, ok := (ps.Primary.Primitive).(*aeadsubtle.XChaCha20Poly1305); ok {
+		ivSize = chacha20poly1305.NonceSizeX
+	}
+
+	return ivSize
+}
+
+// Decrypt will decrypt cipher using the implementation's corresponding encryption key
+func (t *Crypto) Decrypt(cipher, nonce, aad []byte, kh *keyset.Handle) ([]byte, error) {
+	a, err := aead.New(kh)
+	if err != nil {
+		return nil, err
+	}
+
+	ps, err := kh.Primitives()
+	if err != nil {
+		return nil, err
+	}
+
+	ct := make([]byte, 0, len(ps.Primary.Prefix)+len(nonce))
+	ct = append(ct, ps.Primary.Prefix...)
+	ct = append(ct, nonce...)
+	ct = append(ct, cipher...)
```

Comment:
looks like you missed to count cipher length as well e.g `len(ps.Primary.Prefix)+len(nonce)+len(cipher)`

Reply:
good point, although append will expand the array's capacity on the fly.. I'll add it anyway